rpc/types.go

```diff
@@ -98,6 +98,22 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
 	return nil
 }
 
+// MarshalText implements encoding.TextMarshaler. It marshals:
+// - "latest", "earliest" or "pending" as strings
+// - other numbers as hex
+func (bn BlockNumber) MarshalText() ([]byte, error) {
```

Comment:
Essentially, BlockNumber.UnmarshalJSON (rpc/types.go:72) and UnmarshalText both require BlockNumbers to be given as a hex string (or one of the other cases like earliest, pending, latest). But it won't support unmarshaling from an int64.

Without a custom BlockNumber.MarshalText(), it will be marshaled to its int64 representation, and cannot be unmarshaled later. Blocknumbers are part of things that are marshaled and passed around in many places in the codebase but it seems we never tried to unmarshal a previously marshaled blocknumber.

Reply:
I'll test and reproduce this issue to confirm.

Reply:
Ah ok, I track :+1: